File: LeetCode/grind75/206 - Reverse Linked List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next


class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        current = head
        root = None

        while current:
            next_node = current.next
            current.next = root
            root = current
            current = next_node

        return root

File: LeetCode/206 - Reverse Linked List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        ListNode *temp = NULL, *previous = NULL;
        while (head) {
            temp = head->next;
            head->next = previous;
            previous = head;
            head = temp;
        }
        return head = previous;
    }
};